---
title: "WECARE only PCA"
author: "AL"
date: "04/10/2020"
output: 
  html_document:
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: console
---

# Summary

PCA for wecare-only (b37) wes dataset

<style>
pre{
  overflow-x: auto;
}
pre code{
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r echo=F}

options(width=999)

```

# Start section

```{r}

Sys.time()
gc()

rm(list=ls())
base_folder <- "/rds/project/erf33/rds-erf33-medgen"
project_folder <- file.path(base_folder,"users/alexey/wecare/reanalysis_wo_danish_2020")
data_folder <- file.path(project_folder,"data","s05_pca")
scripts_folder <- file.path(project_folder,"scripts","s05_pca")
setwd(scripts_folder)

library(bigsnpr) # for bed_autoSVD() and bed()
library(bigutilsr) # for prob_dist() and tukey_mc_up() for outlier detection
library(hexbin) # for plotting svd loadings
library(ggplot2)
library(dplyr)

NCORES <- 5 # 6 cores were requested in sintr
#NCORES <- nb_cores() doesnt work on cluster
 
```

# Read bed-bim-fam file-set

```{r}

# Location of bed file
bed_file <- file.path(data_folder,"s05_relatedness_check","wecare_only_biallelic_snps_autosomal_comon_norel.bed")

# Read (=attach?) data
kgp.bed <- bed(bed_file) # bigsnpr::bed
kgp.bed

# Explore kgp.bed
attributes(kgp.bed)
str(kgp.bed)

kgp.bed$bedfile
kgp.bed$address

# Clean-up
rm(bed_file)

```

# Phenotypes

```{r}

# fam file
kgp_fam.df <- kgp.bed$fam
dim(kgp_fam.df)
head(kgp_fam.df)

# phhenotypes from external file
kgp_phenotypes.df <- read.table(file.path(base_folder,"s00_phenotypes","igsr-1000_genomes_phase_3_release.tsv"), 
                                header=T, sep="\t", quote="")
dim(kgp_phenotypes.df)
head(kgp_phenotypes.df)

# Merge fam-file and phenotypes from external file 
phenotypes.df <- left_join(kgp_fam.df, kgp_phenotypes.df, by=c("sample.ID"="Sample.name"))
dim(phenotypes.df)
head(phenotypes.df)

# Make sure that dplyr::left_joint hasnt changed the order of samples
sum(phenotypes.df$sample.ID != kgp_fam.df$sample.ID)

```

# Variants

```{r}

# map file
kgp_map.df <- kgp.bed$map
dim(kgp_map.df)
head(kgp_map.df)

# make simple counts 
kgp_maf.df <- bed_MAF(kgp.bed)
dim(kgp_maf.df)
head(kgp_maf.df)

# merge map file with the counts
variants.df <- cbind(kgp_map.df,kgp_maf.df)
dim(variants.df)
head(variants.df)

# Variants with AF(ref) < AF(alt)
x <- variants.df$ac != variants.df$mac
sum(x)
head(variants.df[x,])

# Clean-up
rm(x)

```

# Calculate PCA

Takes care about LD etc.  
See ?plot.big_SVD for plotting svd objets.  

```{r}

# bigsnpr::bed_autoSVD, Default k = 10
kgp.svd <- bed_autoSVD(kgp.bed, k=20, ncores = NCORES) 

# Variants not in LD (detected by clumping during autoSVD)
vars_not_in_LD <- attr(kgp.svd, "subset")
length(vars_not_in_LD)

#attributes(kgp.svd)
str(kgp.svd)

# Eigenvalues
length(kgp.svd$d)
kgp.svd$d
plot(kgp.svd) # default type="screeplot" see ?plot.big_SVD  

# Eigenvectors
dim(kgp.svd$u)
head(kgp.svd$u)

# PCA summary (for PCs from 1 to 20)
plot(kgp.svd,type = "scores",scores=1:20,coeff=0.4)

# Loadings
dim(kgp.svd$v)
head(kgp.svd$v)

# Loadings summary (for PCs from 1 to 20)
plot(kgp.svd,type="loadings",loadings=1:10,coeff=0.4)

```

# PCA plots for 5 super-populations

```{r}

plot(kgp.svd, type = "scores") +
     aes(color = phenotypes.df$Superpopulation.code) +
     labs(title = NULL, color = "Population")

plot(kgp.svd, type = "scores", scores=3:4) +
     aes(color = phenotypes.df$Superpopulation.code) +
     labs(title = NULL, color = "Population")

plot(kgp.svd, type = "scores", scores=5:6) +
     aes(color = phenotypes.df$Superpopulation.code) +
     labs(title = NULL, color = "Population")

plot(kgp.svd, type = "scores", scores=7:8) +
     aes(color = phenotypes.df$Superpopulation.code) +
     labs(title = NULL, color = "Population")

plot(kgp.svd, type = "scores", scores=9:10) +
     aes(color = phenotypes.df$Superpopulation.code) +
     labs(title = NULL, color = "Population")

plot(kgp.svd, type = "scores", scores=11:12) +
     aes(color = phenotypes.df$Superpopulation.code) +
     labs(title = NULL, color = "Population")

plot(kgp.svd, type = "scores", scores=13:14) +
     aes(color = phenotypes.df$Superpopulation.code) +
     labs(title = NULL, color = "Population")

plot(kgp.svd, type = "scores", scores=15:16) +
     aes(color = phenotypes.df$Superpopulation.code) +
     labs(title = NULL, color = "Population")

plot(kgp.svd, type = "scores", scores=17:18) +
     aes(color = phenotypes.df$Superpopulation.code) +
     labs(title = NULL, color = "Population")

plot(kgp.svd, type = "scores", scores=19:20) +
     aes(color = phenotypes.df$Superpopulation.code) +
     labs(title = NULL, color = "Population")

```

# PCA plots for 26 kgp populations

```{r}

plot(kgp.svd, type = "scores") +
     aes(color = phenotypes.df$Population.code) +
     labs(title = NULL, color = "Population")

plot(kgp.svd, type = "scores", scores=3:4) +
     aes(color = phenotypes.df$Population.code) +
     labs(title = NULL, color = "Population")

plot(kgp.svd, type = "scores", scores=5:6) +
     aes(color = phenotypes.df$Population.code) +
     labs(title = NULL, color = "Population")

plot(kgp.svd, type = "scores", scores=7:8) +
     aes(color = phenotypes.df$Population.code) +
     labs(title = NULL, color = "Population")

plot(kgp.svd, type = "scores", scores=9:10) +
     aes(color = phenotypes.df$Population.code) +
     labs(title = NULL, color = "Population")

plot(kgp.svd, type = "scores", scores=11:12) +
     aes(color = phenotypes.df$Population.code) +
     labs(title = NULL, color = "Population")

plot(kgp.svd, type = "scores", scores=13:14) +
     aes(color = phenotypes.df$Population.code) +
     labs(title = NULL, color = "Population")

plot(kgp.svd, type = "scores", scores=15:16) +
     aes(color = phenotypes.df$Population.code) +
     labs(title = NULL, color = "Population")

plot(kgp.svd, type = "scores", scores=17:18) +
     aes(color = phenotypes.df$Population.code) +
     labs(title = NULL, color = "Population")

plot(kgp.svd, type = "scores", scores=19:20) +
     aes(color = phenotypes.df$Population.code) +
     labs(title = NULL, color = "Population")

```

# Selecting outliers 

```{r}

# Eigenvectors
U <- kgp.svd$u

# Measure(s) of "outlieness"  
prob <- prob_dist(U, ncores=NCORES) # bigutilsr::prob_dist
S <- prob$dist.self / sqrt(prob$dist.nn) # alternatively: test$dist.self / test$dist.nn ?
tukey_threshold <- tukey_mc_up(S) # bigutilsr::tukey_mc_up

# Histogram by "outlieness" score
ggplot() +
  geom_histogram(aes(S), color = "black", fill = "blue", alpha = 0.3) +
  theme_bigstatsr() +
  geom_vline(xintercept=tukey_threshold, colour="red") +
  labs(x = "Statistic of outlierness (S)", y = "Frequency (sqrt-scale)")

# Indices for outliers/non-outliers (for convinience)
# Indices are numeric (not logical): rbigparallelr::rows_along(x): seq_len(nrow(x))
# The threshold of 0.9 has been selected manually from the hhistogram
samples_wo_outliers <- rows_along(kgp.bed)[S < tukey_threshold]
outliers <- rows_along(kgp.bed)[S >= tukey_threshold]
length(outliers)

# Show outliers
phenotypes.df[outliers,]

# Save the list of outliers
file_name=file.path(base_folder,"s07_pca","outliers.txt")
write.table(phenotypes.df[outliers,c("family.ID","sample.ID")],
            file_name,quote=F,sep="\t",row.names=F,col.names=F)

# Clean-up
rm(U,prob,file_name)

```

# Visualising "outlieness" on scatterplots

```{r}

# Scatterplot with "outlieness" score colour-coded
plot(kgp.svd, 
     type = "scores", 
     scores = c(1,2), 
     coeff = 0.6) +
  aes(color = S) + 
  scale_colour_viridis_c()

# Scatterplot with outliers colour-coded
plot(kgp.svd, 
     type = "scores", 
     scores = c(1,2), 
     coeff = 0.6) +
  aes(color = S > tukey_threshold)

# Clean-up
rm(S, tukey_threshold)

```

# PCA w/o outliers

```{r}

# Repeating bed_autoSVD w/o outliers usisng already pre-selected variants:  
# To repeat selecting the varaints, which are not in LD again (i.e. repeat the clumping): 
# ind.col= ars_not_in_LD and thr.r2=NA could be omitted.
kgp.no.outliers.svd <- bed_autoSVD(kgp.bed,
                        ind.row = samples_wo_outliers,
                        ind.col = vars_not_in_LD, 
                        thr.r2 = NA,
                        ncores = NCORES)

# Eigenvalues summary
plot(kgp.no.outliers.svd)

# Eigenvectors summary
plot(kgp.no.outliers.svd, 
     type = "scores", 
     scores = 1:10, 
     coeff = 0.4)

# Loadings summary
plot(kgp.no.outliers.svd, 
     type = "loadings", 
     loadings = 1:10, 
     coeff = 0.4)

```

# Projecting the outliers to PC-s calculated w/o outliers

A simple projecting procedure would be multiplying the genotypes by the corresponding PC loadings.  
However, this simpe procdure would be prone for "shrinkage" in hight PCs (e.g. PCs > 10).  
The augmented algorithm (OADP) is used to avoid shrinkage.  

```{r}

# Calculate the "projection" - whatever it is - for the outliers ...
proj <- bed_projectSelfPCA(kgp.no.outliers.svd, 
                           kgp.bed,
                           ind.row = outliers,
                           ncores = 1) # for a few individuals

# Make NA matrix to allocate PCs for all samples: not-outliers and projected-outliers
pc.mx <- matrix(NA, nrow(kgp.bed), ncol(kgp.no.outliers.svd$u))

# A special case of using predict.big_SVD to copy
# scored available from kgp.no.outliers.svd to the new matrix
# (i.e. omitting the outliers rows)
pc.mx[samples_wo_outliers, ] <- predict(kgp.no.outliers.svd)

# Add the "projected" scores for outlying indivoduals
pc.mx[outliers, ] <- proj$OADP_proj

# Plot projeted outliers
plot(pc.mx[samples_wo_outliers, 1:2], pch=1, xlab="PC1", ylab="PC2", 
     main="PC1 vs PC2 recalulated w/o outliers,\nthen outliers projected (red)")
points(pc.mx[outliers, 1:2], pch=19, col = "red")

# Clean-up
rm(proj, pc.mx)

```

# Save results

```{r}

#save.image(file.path(base_folder,"s07_pca","pca.RData"))

```

# End section

```{r}

ls()
Sys.time()
gc()

```
