---
title: "Projecting WECARE on 1KGP"
author: "AL"
date: "06/10/2020"
output:
  html_document:
    toc: true
    toc_float: true
editor_options:
  chunk_output_type: console
---

# Summary

<style>
pre{
  overflow-x: auto;
}
pre code{
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r echo=F}

options(width=999)

```

# Start section

```{r}

Sys.time()
gc()

rm(list=ls())
base_folder <- "/rds/project/erf33/rds-erf33-medgen"
project_folder <- file.path(base_folder,"users/alexey/wecare/reanalysis_wo_danish_2020")
data_folder <- file.path(project_folder,"data","s05_pca")
scripts_folder <- file.path(project_folder,"scripts","s05_pca")
setwd(scripts_folder)

library(bigsnpr) # for bed_autoSVD() and bed()
library(bigutilsr) # for prob_dist() and tukey_mc_up() for outlier detection
library(hexbin) # for plotting svd loadings
library(ggplot2)
library(dplyr)

NCORES <- 5 # 6 cores were requested in sintr
#NCORES <- nb_cores() doesnt work on cluster

```

# Read wecare data

```{r}

# Location of bed file
bed_file <- file.path(data_folder,"s05_relatedness_check","wecare_biallelic_snps_autosomal_common_norel.bed")

# Read (=attach?) data
wecare.bed <- bed(bed_file) # bigsnpr::bed
wecare.bed

# Clean-up
rm(bed_file)

```

# Read wecare phenotypes

```{r}

data_file <- file.path(data_folder,"s12_phenotypes_with_PCs.RData")

load(data_file)
dim(phenotypes.df)

rm(data_file)

```

# Read b37 1KGP data

```{r}

# Location of bed file
bed_file <- file.path(base_folder,"resources","1kgp","b37","s08_remove_outliers","kgp_b37_biallelic_snps_autosomal_common_non_related_no_outliers.bed")

# Read (=attach?) data
kgp.bed <- bed(bed_file) # bigsnpr::bed
kgp.bed

# Clean-up
rm(bed_file)

```

# Load population descriptions for 1KGP

```{r}

data_file <- file.path(base_folder,"resources","1kgp","b37","s00_phenotypes","igsr-1000_genomes_phase_3_release.tsv")

kgp_pop.df <- read.table(data_file, header=T, sep = "\t")
dim(kgp_pop.df)

rm(data_file)

```

# Check variants overlap between 1kgp and cclg

```{r}

wecare_map.df <- wecare.bed$map
dim(wecare_map.df)

kgp_map.df <- kgp.bed$map
dim(kgp_map.df)

x <- inner_join(kgp_map.df, wecare_map.df, by=c("chromosome","physical.pos"))

dim(x)
head(x)

rm(x)

```

# Project Wecare to 1KGP

A simple projecting procedure would be multiplying the genotypes by the corresponding PC loadings.  
However, the augmented algorithm (OADP) is used to avoid shrinkage (the shrinkage is becoming evident in PCs > 5).

```{r}

# Calculate the "projection"
proj <- bed_projectPCA(kgp.bed, wecare.bed, ncores = NCORES)

# Explore proj ...
str(proj)
# proj$obj.svd.ref # big_SVD object computed from reference data.

```

```{r}

# Get 1KGP PCs
kgen_pcs.mx <- predict(proj$obj.svd.ref)

# Get wecare projections
wecare_pcs.mx <- proj$OADP_proj 
dim(wecare_pcs.mx)
# OADP projection of new data into space of reference PCA
# proj$obj.svd.ref # big_SVD object computed from reference data.

# Make plot
x <- cbind(kgen_pcs.mx[, 1:2],"grey")
y <- cbind(wecare_pcs.mx[, 1:2],"red")
z <- rbind(x,y)

plot(z[,1:2],col=z[,3], xlab="PC1", ylab="PC2",
     main="Projecting WECARE to 1KGP")
text(50,-10,"AFR")
text(-27,-37,"EAS")
text(-20,0,"AMR")
text(-12,10,"SAS")
text(-12,32,"EUR")
# Clean-up
rm(x,y,z)

```

# End section

```{r}

ls()
Sys.time()
gc()

```
